indexed_gzip/zran.c

```diff
@@ -2652,13 +2659,17 @@ int zran_import_index(zran_index_t *index,
     zran_point_t *new_list = NULL;
 
     /* Check if file is read only. */
-    if (!is_readonly(fd)) goto fail;
+    if (!is_readonly(fd, f)) goto fail;
 
     /* Read magic bytes, and check for file errors and EOF. */
-    f_ret = fread(magic_bytes, sizeof(magic_bytes), 1, fd);
-
-    if (feof(fd))   goto eof;
-    if (ferror(fd)) goto read_error;
+    f_ret = fread_(magic_bytes, sizeof(magic_bytes), 1, fd, f);
+
+    // We can't check for EOF if we only have a Python file-like object (f) here,
+    // as we haven't read compressed_size yet. Thus, we only call the feof
```

Comment:
Good point. I think that rather than changing the behavior to call `len` in `_feof_python`, I'll use the technique used elsewhere in the code (seeking to the end and using `tell`).

Reply:
Actually, @pauldmccarthy , I looked further into this and realized that `feof` can't be properly emulated by just calling `.seek()` and checking to see if the position is at the file size; doing this caused some existing functionality to break. This is because `feof()` checks an EOF indicator that is _only set after one attempts to read a file_.

Thus, I've changed the code to actually store an EOF indicator on the file object (called `__indexed_gzip_eof`), which is only set when the file is read and it returns nothing. I guess this is kind of ugly, but I think it's the best way to emulate the EOF functionality. It works because whenever we call `feof_` in the code, it's always preceded by a call to `fread_`. Let me know what you think.

Reply:
Actually, this EOF indicator approach doesn't work on Python 2.7 because I can't add new attributes to built-in objects there.

I've changed the logic to just check if `f_read == 0` to determine if a Python file-like object is at EOF.

Reply:
This seems like a reasonable solution. Preserving support for python 2 is not a major priority for me, but I think some bio communities may still be coupled to python 2, so it's probably good to preserve support if not too much hassle.

I'm happy with this solution, but maybe a future clean-up could move the check out to the calling code - e.g.:
```
ret = fread_(buf, 4, 4, fd, f);
if (feof_(fd, f) || !ret) goto eof;
```

Not really any better than what you've got here though.